fix building on macOS

kambala-decapitator opened the pull request:
I had the following errors with Apple clang 13 (Xcode 13.4.1):

```
attr.c:560:5: error: implicitly declaring library function 'free' with type 'void (void *)' [-Werror,-Wimplicit-function-declaration]
    free(bad_dir_cache[best].path);
    ^
attr.c:560:5: note: include the header <stdlib.h> or explicitly provide a declaration for 'free'
bison -y -d ./exports.y
attr.c:577:6: error: implicitly declaring library function 'strcmp' with type 'int (const char *, const char *)' [-Werror,-Wimplicit-function-declaration]
        if (strcmp(path, bad_dir_cache[i].path) == 0) {
            ^
attr.c:577:6: note: include the header <string.h> or explicitly provide a declaration for 'strcmp'
attr.c:585:23: error: implicitly declaring library function 'strdup' with type 'char *(const char *)' [-Werror,-Wimplicit-function-declaration]
    new_entry->path = strdup(path);
                      ^
attr.c:585:23: note: include the header <string.h> or explicitly provide a declaration for 'strdup'
```

CendioOssman commented:
Thanks. I tried setting up automated builds on macOS to catch issues like this, but unfortunately homebrew doesn't have TI-RPC for macOS.

What steps did you need to do to get things building?

kambala-decapitator commented:
nothing special was required, I've downloaded the source from release and ran
```
chmod +x configure
./configure
make -j6
```

configure output:
```
checking for a BSD-compatible install... /usr/bin/install -c
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for ar... ar
checking for flex... flex
checking lex output file root... lex.yy
checking lex library... -ll
checking whether yytext is a pointer... yes
checking for bison... bison -y
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for library containing xdr_int... none required
checking for library containing socket... none required
checking for library containing inet_aton... none required
checking for mntent.h... no
checking for stdint.h... yes
checking for sys/mnttab.h... no
checking for sys/mount.h... yes
checking for sys/vmount.h... no
checking for rpc/svc_soc.h... no
checking for linux/ext2_fs.h... no
checking for int32... no
checking for uint32... no
checking for int64... no
checking for uint64... no
checking for struct stat.st_gen... yes
checking for struct __rpc_svcxprt.xp_fd... no
checking for xdr_int... yes
checking for xdr_u_int... yes
checking for xdr_int32... no
checking for xdr_int32_t... yes
checking for xdr_uint32... no
checking for xdr_uint32_t... no
checking for xdr_u_int32_t... yes
checking for xdr_uint64... no
checking for xdr_uint64_t... no
checking for xdr_u_int64_t... yes
checking for svc_getreq_poll... no
checking for statvfs... yes
checking for seteuid... yes
checking for setegid... yes
checking for setresuid... no
checking for setresgid... no
checking for vsyslog... yes
checking for lchown... yes
checking for setgroups... yes
checking for svc_tli_create... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating Config/Makefile
config.status: creating Extras/Makefile
config.status: creating config.h
```

btw there was some strange issue when unpacking the archive:
```
Would you like to replace the existing file:
  Path:     ./unfs3-0.9.23/contrib/nfsotpclient/mountclient/__init__.py
  Size:     119 bytes (1 KiB)
  Modified: 2022-10-06 17:35:58
with the file from archive:
  Path:     unfs3-0.9.23/contrib/nfsotpclient/mountclient/__init__.py
  Size:     0 bytes
  Modified: 2022-10-06 17:35:58
? (Y)es / (N)o / (A)lways / (S)kip all / A(u)to rename all / (Q)uit? n
``` 
and same with all files under `contrib/nfsotpclient/mountclient`, I didn't overwrite them with the 0 bytes version.

kambala-decapitator commented:
not sure if important, but I have installed `userspace-rcu` from brew - I was trying to build https://github.com/nfs-ganesha/nfs-ganesha (which failed when trying to build bundled copy of nTIRPC and I didn't bother investigating)

CendioOssman commented:
Ah. I'm afraid that is not the current release of unfs3. Please try the 0.10.0 one. Which unfortunately will require TI-RPC. And I don't know how to get that on to macOS.

GitHub for some reason shows the wrong release as "latest". I'm trying to find a solution to it, but GitHub's support have been unable to resolve it. :/

I'll have a look at that archive issue.

kambala-decapitator commented:
yes, the latest version fails with
```
configure: error: Package requirements (libtirpc) were not met:

No package 'libtirpc' found
```

> GitHub for some reason shows the wrong release as "latest". I'm trying to find a solution to it, but GitHub's support have been unable to resolve it. :/

does Release - Edit - "mark this release as the latest" checkbox not work?

> I'll have a look at that archive issue.

FYI same unpack issue with the latest version

kambala-decapitator commented:
just tried building https://sourceforge.net/projects/libtirpc/ on macOS: solved some build errors, but kinda lazy to continue.

what I changed so far:

1. `#if defined(__linux__)` in `tirpc/reentrant.h` needs to be extended to check for darwin (for test I simply used `#if 1`)
2. `#if defined(__UCLIBC__) && !defined(__UCLIBC_HAS_RPC__) || !defined(__GLIBC__)` in `tirpc/rpc/rpcent.h` needs to be extended to ignore darwin (for test I simply used `#if 0`)
3. `SOL_IPV6` in `src/rpc_generic.c` needs to be replaced with `IPPROTO_IPV6` on darwin/bsd

further errors I received which also need darwin/bsd counterparts:

```
svc_dg.c:657:25: error: use of undeclared identifier 'SOL_IP'
                (void) setsockopt(fd, SOL_IP, IP_PKTINFO, &val, sizeof(val));
                                      ^
svc_dg.c:661:25: error: use of undeclared identifier 'SOL_IPV6'
                (void) setsockopt(fd, SOL_IPV6, IPV6_RECVPKTINFO, &val, sizeof(val));
                                      ^
svc_dg.c:661:35: error: use of undeclared identifier 'IPV6_RECVPKTINFO'
                (void) setsockopt(fd, SOL_IPV6, IPV6_RECVPKTINFO, &val, sizeof(val));
                                                ^
svc_dg.c:689:27: error: use of undeclared identifier 'SOL_IP'
                if (cmsg->cmsg_level != SOL_IP
                                        ^
svc_dg.c:703:27: error: use of undeclared identifier 'SOL_IPV6'
                if (cmsg->cmsg_level != SOL_IPV6
                                        ^
svc_dg.c:704:26: error: use of undeclared identifier 'IPV6_PKTINFO'
                 || cmsg->cmsg_type != IPV6_PKTINFO
```

CendioOssman commented:
> does Release - Edit - "mark this release as the latest" checkbox not work?

Huh. That's new... It seems to partially solve the issue. It's now tagged as "latest" at least. But the ordering is still off. :/

kambala-decapitator commented:
> But the ordering is still off. :/

looks correct to me:

![Screenshot 2022-11-07 at 14 56 05](https://user-images.githubusercontent.com/1557784/200304427-5f3bb107-b8b5-4738-8f60-9bc8f7d0c513.png)

![Screenshot 2022-11-07 at 14 55 57](https://user-images.githubusercontent.com/1557784/200304420-88d7aac1-5c25-4a55-aea9-2838f814bd66.png)

CendioOssman commented:
I messed around with the tags a bit more which seems to have fixed the ordering.

The bug in the archives should be gone now as well.

CendioOssman commented:
I've merged your fix now, since it was correct for `master` as well. Thanks!

Unfortunately we cannot do much about TI-RPC on our end. :/
Have you tried poking them about macOS support?

kambala-decapitator commented:
don't have permissions to open a ticket there, opened a discussion instead: https://sourceforge.net/p/libtirpc/discussion/637321/thread/30b9265f2f/

although checking other tickets/threads I doubt there'd be any help

gahfy commented:
@kambala-decapitator hope this will help, this patch allowing you to install `libtirpc-1.3.3` on macOS. You will have to disable GSSAPI (`--disable-gssapi` in `configure` script)

```
diff --git a/src/Makefile.in b/src/Makefile.in
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -90,7 +90,7 @@
 build_triplet = @build@
 host_triplet = @host@
 @AUTHDES_TRUE@am__append_1 = auth_des.c  authdes_prot.c  des_crypt.c  des_impl.c  des_soft.c  svc_auth_des.c
-@SYMVERS_TRUE@am__append_2 = -Wl,--version-script=$(srcdir)/libtirpc.map
+@SYMVERS_TRUE@am__append_2 = -Wl
 @GSS_TRUE@am__append_3 = auth_gss.c authgss_prot.c svc_auth_gss.c \
 @GSS_TRUE@			   rpc_gss_utils.c
 
diff --git a/src/getpeereid.c b/src/getpeereid.c
--- a/src/getpeereid.c
+++ b/src/getpeereid.c
@@ -24,7 +24,7 @@
  * SUCH DAMAGE.
  */
 
-
+#ifndef __APPLE__
 #include <sys/param.h>
 #include <sys/socket.h>
 #include <sys/un.h>
@@ -49,3 +49,4 @@
 	*egid = uc.gid;
 	return (0);
  }
+#endif
diff --git a/src/rpc_com.h b/src/rpc_com.h
--- a/src/rpc_com.h
+++ b/src/rpc_com.h
@@ -42,6 +42,11 @@
 
 #include <rpc/rpc_com.h>
 
+#ifdef __APPLE__
+    #define SOL_IPV6 IPPROTO_IPV6
+    #define SOL_IP IPPROTO_IP
+#endif
+
 #ifdef __cplusplus
 extern "C" {
 #endif
diff --git a/src/svc_dg.c b/src/svc_dg.c
--- a/src/svc_dg.c
+++ b/src/svc_dg.c
@@ -37,6 +37,7 @@
  *
  * Does some caching in the hopes of achieving execute-at-most-once semantics.
  */
+#define __APPLE_USE_RFC_3542
 #include <pthread.h>
 #include <reentrant.h>
 #include <sys/types.h>
diff --git a/src/svc_raw.c b/src/svc_raw.c
--- a/src/svc_raw.c
+++ b/src/svc_raw.c
@@ -43,6 +43,7 @@
 #include <sys/types.h>
 #include <rpc/raw.h>
 #include <stdlib.h>
+#include <string.h>
 
 #ifndef UDPMSGSIZE
 #define	UDPMSGSIZE 8800
diff --git a/src/xdr_float.c b/src/xdr_float.c
--- a/src/xdr_float.c
+++ b/src/xdr_float.c
@@ -83,7 +83,11 @@
 };
 #else
 
+#ifndef __APPLE__
 #include <endian.h>
+#else
+#include <machine/endian.h>
+#endif
 #define IEEEFP
 
 #endif /* vax */
diff --git a/tirpc/reentrant.h b/tirpc/reentrant.h
--- a/tirpc/reentrant.h
+++ b/tirpc/reentrant.h
@@ -36,7 +36,7 @@
  * These definitions are only guaranteed to be valid on Linux. 
  */
 
-#if defined(__linux__)
+#if defined(__linux__) || defined(__APPLE__)
 
 #include <pthread.h>
 
diff --git a/tirpc/rpc/rpcent.h b/tirpc/rpc/rpcent.h
--- a/tirpc/rpc/rpcent.h
+++ b/tirpc/rpc/rpcent.h
@@ -50,7 +50,7 @@
 
 /* These are defined in /usr/include/rpc/netdb.h, unless we are using
    the C library without RPC support. */
-#if defined(__UCLIBC__) && !defined(__UCLIBC_HAS_RPC__) || !defined(__GLIBC__)
+#if defined(__UCLIBC__) && !defined(__UCLIBC_HAS_RPC__) || !defined(__GLIBC__) && !defined(__APPLE__)
 struct rpcent {
 	char	*r_name;	/* name of server for this rpc program */
 	char	**r_aliases;	/* alias list */
```

Note : The only error that prevents from building with GSSAPI is the undefined function `gss_pname_to_uid()` call in `src/svc_auth_gss.c`. Unfortunately, when I try to find this declaration in my system with the following command, I got no result.

```shell
sudo find / -name gssapi.h 2>/dev/null | xargs -I '{}' grep "gss_pname_to_uid" '{}'
```

Feel free to tell me if you find a way to solve it

kambala-decapitator commented:
@gahfy great work!

> gss_pname_to_uid

no idea, sorry. Seems that Apple simply doesn't provide such API.
